packages/node-runtime-worker-thread/src/serializer.ts

```diff
@@ -0,0 +1,145 @@
+  return Object.assign(new Error(), err);
+}
+
+export enum SerializedResultTypes {
+  SerializedErrorResult = 'SerializedErrorResult',
+  InspectResult = 'InspectResult',
+  SerializedShellApiResult = 'SerializedShellApiResult'
+}
+
+type SerializedShellApiResult = {
+  origType: string;
+  serializedValue: Document;
+  nonEnumPropertyDescriptors: Record<string, PropertyDescriptor>;
+};
+
+export function serializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  // Type `null` indicates anything that is not returned by shell-api, any
+  // usual javascript value returned from evaluation
+  if (result.type === null) {
+    // Errors get special treatment to achieve better serialization
+    if (isError(result.rawValue)) {
+      result.type = SerializedResultTypes.SerializedErrorResult;
+      result.printable = serializeError(result.rawValue);
+    } else if (!isPrimitive(result.rawValue)) {
+      // For everything else that is not a primitive value there is just too
+      // many possible combinations of data types to distinguish between them,
+      // if we don't know the type and it's not an error, let's inspect and
+      // return an inspection result
+      result.type = SerializedResultTypes.InspectResult;
+      result.printable = inspect(result.rawValue);
+    }
+  } else if (!isPrimitive(result.printable)) {
+    // If type is present we are dealing with shell-api return value. In most
+    // cases printable values of shell-api are primitive or serializable as
+    // (E)JSON, but they also might have some non-enumerable stuff (looking at
+    // you, Cursor!) that we need to serialize additionally, otherwiser those
+    // params will be lost in the ether causing issues down the way
+    const origType = result.type;
+    result.type = SerializedResultTypes.SerializedShellApiResult;
+    result.printable = {
+      origType,
+      serializedValue: EJSON.serialize(result.printable),
+      nonEnumPropertyDescriptors: Object.fromEntries(
+        Object.entries(
+          Object.getOwnPropertyDescriptors(result.printable)
+        ).filter(([, descriptor]) => !descriptor.enumerable)
+      )
+    };
+  }
+
+  // `rawValue` can't be serialized "by design", we don't really care about it
+  // for the worker thread runtime so as an easy workaround we will just remove
+  // it completely from the result
+  delete result.rawValue;
+
+  return result;
+}
+
+export function deserializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  if (result.type === SerializedResultTypes.SerializedErrorResult) {
+    result.type = null;
+    result.printable = deserializeError(result.printable);
+  }
+
+  if (result.type === SerializedResultTypes.SerializedShellApiResult) {
+    const value: SerializedShellApiResult = result.printable;
+    const printable = EJSON.deserialize(value.serializedValue);
+
+    // Primitives should not end up here ever, but we need to convince TS that
```

Comment:
Again i think it would be clearer if we take care of this `if` case right in the beginning of the function, keep it out of our way and mind in the rest of the function.

Reply:
It's the `EJSON.deserialize` return type (`type SerializableTypes = Document | Array<JSONPrimitive | Document> | JSONPrimitive`) that asserts primitive types on the `printable`, so this check has to be here